solarkraft/src/state/value.ts

```diff
@@ -11,133 +53,125 @@ interface Value { }
  * 
  * Example: 2u32 would be represented as { val: 2, type: "u32" }, whereas 2i32 would be { val: 2, type: "i32" }
 */
-abstract class IntValue implements Value {
+export type IntValue = {
     val: bigint
-    type: string
+} & (IntValue_u32 | IntValue_i32 | IntValue_u64 | IntValue_i64 | IntValue_u128 | IntValue_i128)
 
-    constructor(v: bigint) {
-        this.val = v
-        if (!this.isValid()) {
-            throw new RangeError(`${v} lies outside the ${this.type} range.`)
-        }
-    }
 
-    abstract isValid(): boolean
+function mkInt(type: string, val: bigint) {
+    const obj = { type: type, val: val } as Value
```

Comment:
this does not seem type-safe to me.

```suggestion
function mkInt(type: "u32" | "u64" | ... | "u128", val: bigint) {
    const obj = { type: type, val: val }
```